provision/docker/provisioner.go

```diff
@@ -1008,6 +1008,19 @@ func (p *dockerProvisioner) ListNodes(addressFilter []string) ([]provision.Node,
 	return result, nil
 }
 
+func (p *dockerProvisioner) ListNodesByFilter(filter map[string]string) ([]provision.Node, error) {
+	nodes, err := p.Cluster().UnfilteredNodesForMetadata(filter)
+	if err != nil {
+		return nil, err
+	}
+	result := make([]provision.Node, 0, len(nodes))
```

Comment:
You can safely allocate the `result` slice with the `len(nodes)` length.

```suggestion
	result := make([]provision.Node, len(nodes))
```